src/tensors/tensor.jl

```diff
@@ -70,89 +67,215 @@ storagetype(::Type{<:TensorMap{T, S, N₁, N₂, A}}) where {T, S, N₁, N₂, A
 dim(t::TensorMap) = length(t.data)
 
 # General TensorMap constructors
-#--------------------------------
+# ==============================
+
+# INTERNAL: utility type alias that makes constructors also work for type aliases that specify
+# different storage types. (i.e. CuTensorMap = TensorMapWithStorage{T, CuVector{T}, ...})
+const TensorMapWithStorage{T, A <: DenseVector{T}, S, N₁, N₂} = TensorMap{T, S, N₁, N₂, A}
+const TensorWithStorage{T, A <: DenseVector{T}, S, N} = Tensor{T, S, N, A}
+
 # undef constructors
+# ------------------
+# - dispatch start with TensorMap{T}
+# - select A and map to TensorMap{T, S, N₁, N₂, A} where {S, N₁, N₂}
+# - select S, N₁, N₂ and map to TensorMap{T, S, N₁, N₂, A}
 """
-    TensorMap{T}(undef, codomain::ProductSpace{S,N₁}, domain::ProductSpace{S,N₂})
-                where {T,S,N₁,N₂}
+    TensorMap{T}(undef, codomain::ProductSpace{S, N₁}, domain::ProductSpace{S, N₂}) where {T, S, N₁, N₂}
     TensorMap{T}(undef, codomain ← domain)
     TensorMap{T}(undef, domain → codomain)
-    # expert mode: select storage type `A`
-    TensorMap{T,S,N₁,N₂,A}(undef, codomain ← domain)
-    TensorMap{T,S,N₁,N₂,A}(undef, domain → domain)
 
-Construct a `TensorMap` with uninitialized data.
+Construct a `TensorMap` with uninitialized data with elements of type `T`.
 """
-function TensorMap{T}(::UndefInitializer, V::TensorMapSpace{S, N₁, N₂}) where {T, S, N₁, N₂}
-    return TensorMap{T, S, N₁, N₂, Vector{T}}(undef, V)
-end
-function TensorMap{T}(
-        ::UndefInitializer, codomain::TensorSpace{S}, domain::TensorSpace{S}
-    ) where {T, S}
-    return TensorMap{T}(undef, codomain ← domain)
-end
-function Tensor{T}(::UndefInitializer, V::TensorSpace{S}) where {T, S}
-    return TensorMap{T}(undef, V ← one(V))
+TensorMap{T}(::UndefInitializer, V::TensorMapSpace) where {T} =
```

Comment:
So this will allow something like
```julia
TensorMap{Vector{Float64}}(undef, V)
```

Is that intentional?

Reply:
```julia
julia> using TensorKit

julia> V = ComplexSpace(2)
ℂ^2

julia> TensorMap{Vector{Float64}}(undef, V)
ERROR: TypeError: in AbstractTensorMap, in T, expected T<:Number, got Type{Vector{Float64}}
```

That's not actually true, you can't instantiate these types because of the type restrictions in the struct definition, so it will already error before you get to call that method.

Reply:
Ok, this seems like the second time I ask this question today? Or at least a very similar one.

Reply:
I think the way I would think about this is that really you are first constructing the type itself, which is also a julia object, for which the type restrictions of the type definition are checked, and then you are just using that type as a callable and interpreting that as a constructor. In that regards, Julia really doesn't have constructors and this is just a convention that it should return a type that conforms to the callable.